# 2540122041 - Stefan Bondito Giovanno - UTS Text Mining

## Link Video: https://drive.google.com/file/d/1G4abcs6pHYSoa0X3rKVBR8d8BtJTJCXH/view?usp=sharing

# Library

In [ ]:
pip install newspaper3k

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 211.1/211.1 kB 2.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 81.3/81.3 kB 4.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 97.6/97.6 kB 4.6 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7.4/7.4 MB 15.2 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Preparing metadata (setup.py) ... done
  Preparing metadata (setup.py) ... done
  Created wheel for tinysegmenter: filename=tinysegmenter-0.3-py3-none-any.whl size=13538 sha256=78b5f190ea80ec5af77a506dc73055ccf55dd91ed2efae6845b4d50aa09863b0
  Stored in directory: /root/.cache/pip/wheels/c8/d6/6c/384f58df48c00b9a31d638005143b5b3ac62c3d25fb1447f23
  Created wheel for feedfinder2: filename=feedfinder2-0.0.4-py3-none-any.whl size=3340 sha256=c5be24882f83fda2c1ba3d5b9448b9bc75ba369288e2722f17a185b113c2aeb0
  Stored in directory: /root/.cache/pip/wheels/97/02/e7/a1ff1760e12bdbaab0ac824fae5c1bc

In [ ]:
pip install nltk

In [ ]:
from newspaper import Article
import re
import nltk
from nltk.tokenize import word_tokenize

import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, recall_score, precision_score, f1_score, roc_auc_score, log_loss
from sklearn.metrics import classification_report, confusion_matrix

In [ ]:
nltk.download('punkt')

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.


True

#1. News Scraping

Pada bagian ini, link-link dari website berita akan dipecah menjadi kategori-kategorinya, seperti bisnis, politik, olahraga, dan hiburan. Link-link tersebut akan dimasukkan ke dalam sebuah list. Sebuah loop dengan for loop akan digunakan untuk mengakses isi dari list satu per satu dan membuka link tersebut. Beberapa function yang didapatkan dari library **Article** yang diambil lagi dari **newspaper**.

Dengan bantuan function tersebut, kita bisa mendapatkan info-info seperti isi dari artikel, link sumber artikel tersebut yang sekaligus menjelaskan sumber beritanya, dan label dari berita, yang bisa ditentukan karena sudah kita pisahkan secara kategori dari awal. Info-info tersebut akan disimpan ke dalam sebuah dataframe sesuai dengan permintaan soal dengan format tabel yang sama juga. Hasilnya seperti dibawah ini.

## Business

In [ ]:
business_links = [
'https://edition.cnn.com/2024/04/18/business/hyundai-pause-advertising-x-antisemitism/index.html',
'https://apnews.com/article/stocks-markets-shares-interest-d9bd7b0402eb95e4c58d8489ba7d7daa',
'https://www.cnbc.com/2024/04/18/netflix-nflx-earnings-q1-2024.html',
'https://www.cnbc.com/2024/04/19/autonomous-drone-startup-zipline-hits-1-million-deliveries.html',
'https://www.cnbc.com/2024/04/12/nike-ceo-acknowledges-it-went-too-far-in-direct-push.html',
'https://www.cnbc.com/2024/04/07/disney-parks-are-its-top-money-maker-its-spending-to-keep-it-that-way.html',
'https://www.cnbc.com/2024/04/17/superyacht-sales-plunge-as-russian-oligarchs-drop-out-of-the-market.html',
'https://edition.cnn.com/2024/04/19/media/netflix-subscription-numbers/index.html',
'https://apnews.com/article/tiktok-us-ban-sale-china-congress-de12b4d22aa8095e62cb0982a6e62235',
'https://apnews.com/article/stock-market-shares-rates-treasury-c8cc2e5f5ac2b5cd47143dd2ee0920ea',
'https://apnews.com/article/irs-customer-service-tax-season-april-15-dee7ae01dc6061697412a2571d406843',
'https://apnews.com/article/imf-world-economy-inflation-gaza-ukraine-georgieva-6c5a66e7bc6c7ed5fdd017bf57cf2296',
'https://apnews.com/article/tax-season-mental-health-filing-deadline-6b8e9633138c878ae72316c778520d8a',
'https://apnews.com/article/tax-season-filing-child-tax-credit-6893fe49a3db164937e6b7786b865db2',
'https://apnews.com/article/new-york-budget-migrants-housing-marijuana-2456ae469ed0e439a5d25cd96aad84a1',
'https://apnews.com/article/china-gdp-economy-first-quarter-a0743b0982996194c07713c5c701918f',
'https://apnews.com/article/thailand-economy-digital-handout-stimulus-d53f3683c9c9bfd4da33cff8cd2b32e7',
'https://www.cnbc.com/2024/04/19/procter-gamble-pg-q3-2024-earnings.html',
'https://www.cnbc.com/2024/04/19/money-coach-how-i-make-a-living-without-working-all-the-time.html',
'https://www.cnbc.com/2024/04/18/nba-tv-rights-news-on-negotiations-with-espn-warner-bros-discovery.html',
'https://edition.cnn.com/2024/04/24/business/boeing-losses/index.html',
'https://www.cnbc.com/2024/04/24/biogen-biib-earnings-q1-2024-.html',
'https://www.cnbc.com/2024/04/24/boeing-ba-earnings-q1-2024.html',
'https://www.cnbc.com/2024/04/23/walmart-backed-fintech-one-introduces-buy-now-pay-later.html',
'https://www.cnbc.com/2024/04/23/jetblue-jblu-lowers-2024-revenue-outlook.html',
'https://www.cnbc.com/2024/04/22/ex-jetblue-ceo-robin-hayes-to-run-airbus-north-america-.html',
'https://apnews.com/article/stock-markets-rates-sensetime-japan-d87e7b8e038660dd4687eba82537169f',
'https://apnews.com/article/auto-insurance-premiums-inflation-ec0756918e2ddc8e50f698c8b938753c',
'https://apnews.com/article/germany-retirees-pensions-increase-inflation-b720bcd8ce479e6141ea41529d8cc10e',
'https://apnews.com/article/china-gdp-economy-first-quarter-a0743b0982996194c07713c5c701918f'
]


In [ ]:
business_df = pd.DataFrame()

for link in business_links:
  article = Article(link)
  article.download()
  article.parse()

  list_text = []
  list_source = []
  list_label = []

  list_text.append(article.text)
  list_source.append(article.source_url)
  list_label.append('Business')

  temp_df = pd.DataFrame({'Text': list_text, 'Media': list_source, 'Label': list_label})
  business_df = pd.concat([business_df, temp_df], ignore_index=True)


In [ ]:
business_df

Text  \
0   New York CNN —\n\nHyundai said on Thursday it ...   
1   NEW YORK (AP) — The worst week for big technol...   
2   A couple sits in front of a television with th...   
3   Autonomous delivery drone startup Zipline said...   
4   Illuminated trademark of the American athletic...   
5   In this article DIS Follow your favorite stock...   
6   Dazert said he expects new superyacht sales to...   
7   Editor’s Note: A version of this article first...   
8   WASHINGTON (AP) — The House passed legislation...   
9   HONG KONG (AP) — Asian stocks tumbled Friday, ...   
10  WASHINGTON (AP) — On this Tax Day, refunds are...   
11  WASHINGTON (AP) — The head of the Internationa...   
12  NEW YORK (AP) — For many people, tax season is...   
13  NEW YORK (AP) — There are a few tax tips paren...   
14  ALBANY, N.Y. (AP) — New York lawmakers passed ...   
15  HONG KONG (AP) — China’s economy grew faster t...   
16  BANGKOK (AP) — Thailand’s Prime Minister Srett...   
17  In this photo illustration, Pantene and Head &...   
18  Bernadette Joy started her money coaching side...   
19  Los Angeles Lakers forward LeBron James, #23, ...   
20  New York CNN —\n\nBoeing reported a slightly s...   
21  A test tube is seen in front of displayed Biog...   
22  An aerial photo shows Boeing 737 Max airplanes...   
23  In this article COF\n\nAFRM\n\nWMT Follow your...   
24  Silhouette of passenger in front of the JetBlu...   
25  Robin Hayes, chief executive officer of JetBlu...   
26  NEW YORK (AP) — U.S. stocks are drifting lower...   
27  NEW YORK (AP) — Relentlessly rising auto insur...   
28  BERLIN (AP) — Germany’s Cabinet on Wednesday a...   
29  HONG KONG (AP) — China’s economy grew faster t...   

                      Media     Label  
0   https://edition.cnn.com  Business  
1        https://apnews.com  Business  
2      https://www.cnbc.com  Business  
3      https://www.cnbc.com  Business  
4      https://www.cnbc.com  Business  
5      https://www.cnbc.com  Business  
6      https://www.cnbc.com  Business  
7   https://edition.cnn.com  Business  
8        https://apnews.com  Business  
9        https://apnews.com  Business  
10       https://apnews.com  Business  
11       https://apnews.com  Business  
12       https://apnews.com  Business  
13       https://apnews.com  Business  
14       https://apnews.com  Business  
15       https://apnews.com  Business  
16       https://apnews.com  Business  
17     https://www.cnbc.com  Business  
18     https://www.cnbc.com  Business  
19     https://www.cnbc.com  Business  
20  https://edition.cnn.com  Business  
21     https://www.cnbc.com  Business  
22     https://www.cnbc.com  Business  
23     https://www.cnbc.com  Business  
24     https://www.cnbc.com  Business  
25     https://www.cnbc.com  Business  
26       https://apnews.com  Business  
27       https://apnews.com  Business  
28       https://apnews.com  Business  
29       https://apnews.com  Business

## Politics

In [ ]:
politics_links = [
'https://www.cnbc.com/2024/04/18/ukraine-war-live-updates-latest-news-on-russia-and-the-war-in-ukraine.html',
'https://www.cnbc.com/2024/04/16/senator-encourages-drivers-to-drag-gaza-cease-fire-protestors-from-roads.html',
'https://www.cnbc.com/2024/04/16/chinas-inaction-on-iran-has-undermined-its-influence-in-the-middle-east-analyst-says.html',
'https://apnews.com/article/pennsylvania-primary-biden-16-house-fitzpatrick-houck-c5b7c0a05a7dbe9e61b3607767b5f629',
'https://www.cnbc.com/2024/04/11/japanese-pm-kishida-says-the-us-must-play-leading-role-in-the-world.html',
'https://www.cnbc.com/2024/04/03/us-loses-its-spot-to-china-as-southeast-asias-most-favored-ally-survey-finds.html',
'https://www.cnbc.com/2024/03/08/hong-kongs-national-security-draft-has-life-imprisonment-for-treason.html',
'https://www.cnbc.com/2024/02/09/russia-ukraine-what-next-for-the-war-as-zelenskyy-replaces-army-chief.html',
'https://www.cnbc.com/2024/01/31/fbi-shut-down-china-volt-typhoon-hackers-targeting-us-.html',
'https://www.cnbc.com/2024/04/11/bidens-china-tariff-threats-are-more-bark-than-bite-economists-say.html',
'https://www.cnbc.com/2024/03/21/us-rejects-chinas-claim-to-arunachal-pradesh-says-it-is-a-part-of-india.html',
'https://edition.cnn.com/2024/04/20/politics/espionage-us-sailor-pedicini-navy-found-guilty/index.html',
'https://www.cnbc.com/2024/04/20/trump-berates-judge-for-rushed-hush-money-trial-days-before-opening-statements.html',
'https://edition.cnn.com/2024/04/17/politics/fact-check-donald-trump-jimmy-kimmel-al-pacino/index.html?iid=cnn_buildContentRecirc_end_recirc',
'https://apnews.com/article/tiktok-us-ban-sale-china-congress-de12b4d22aa8095e62cb0982a6e62235',
'https://edition.cnn.com/2024/04/20/politics/campaign-finance-reports-trump-legal-fees/index.html',
'https://edition.cnn.com/2024/03/21/politics/trump-bond-deadline/index.html',
'https://edition.cnn.com/2024/04/20/politics/north-carolina-trump-rally/index.html',
'https://edition.cnn.com/2024/02/06/politics/trump-immunity-court-of-appeals/index.html',
'https://www.cnbc.com/2024/04/19/bae-systems-linked-to-deals-in-countries-accused-of-human-rights-abuses-report.html',
'https://www.cnbc.com/2024/04/17/ukraine-war-live-updates-latest-news-on-russia-and-the-war-in-ukraine.html',
'https://www.cnbc.com/2024/04/16/ukraine-war-live-updates-latest-news-on-russia-and-the-war-in-ukraine.html',
'https://www.cnbc.com/2024/04/13/ukraines-military-chief-warns-of-significantly-worsening-battlefield-situation-in-the-east.html',
'https://www.cnbc.com/2024/04/10/russia-ukraine-war-live-updates-from-april-10-2024.html',
'https://www.cnbc.com/2024/04/09/simon-harris-becomes-youngest-ever-irish-prime-minister-pledges-reset.html',
'https://apnews.com/article/don-blankenship-manchin-democrats-west-virginia-senate-f51e9edcedd63204b9fb8a8b27e64a5f',
'https://www.cnbc.com/2024/04/23/ukraine-war-live-updates-latest-news-on-russia-and-the-war-in-ukraine.html',
'https://edition.cnn.com/2024/04/23/politics/us-ukraine-military-aid-package/index.html',
'https://edition.cnn.com/2024/04/23/politics/donald-trump-gag-order-analysis/index.html',
'https://edition.cnn.com/2024/04/10/politics/lara-trump-rnc-2020-election-fraud-claims/index.html'


]


In [ ]:
politics_df = pd.DataFrame()

for link in politics_links:
  article = Article(link)
  article.download()
  article.parse()

  list_text = []
  list_source = []
  list_label = []

  list_text.append(article.text)
  list_source.append(article.source_url)
  list_label.append('Politics')

  temp_df = pd.DataFrame({'Text': list_text, 'Media': list_source, 'Label': list_label})
  politics_df = pd.concat([politics_df, temp_df], ignore_index=True)


In [ ]:
politics_df

Text  \
0   Sweden's Finance Minister Elisabeth Svantesson...   
1   U.S. Senator Tom Cotton (R-AR) speaks during a...   
2   China's restrained silence on the Middle East ...   
3   NEW HOPE, Pa. (AP) — Mark Houck makes a potent...   
4   U.S. Vice President Kamala Harris and U.S. Hou...   
5   A majority of Southeast Asians would align wit...   
6   The flag of Hong Kong flies from a ferry boat ...   
7   Ukrainian President Volodymyr Zelenskyy attend...   
8   FBI Director Christopher Wray testifies before...   
9   U.S. President Joe Biden attends a bilateral m...   
10  A tableau of Arunachal Pradesh state during th...   
11  CNN —\n\nA US sailor who served in Japan was f...   
12  Former US President Donald Trump sits in the c...   
13  Washington CNN —\n\nOn Tuesday, late-night tel...   
14  WASHINGTON (AP) — The House passed legislation...   
15  CNN —\n\nDonald Trump’s leadership PAC spent a...   
16  CNN —\n\nDonald Trump’s three-day deadline to ...   
17  CNN —\n\nFormer President Donald Trump’s rally...   
18  CNN —\n\nDonald Trump is not immune from prose...   
19  LONDON — BAE Systems , the U.K.'s largest defe...   
20  U.S. Treasury Secretary Janet Yellen on Tuesda...   
21  U.S. Treasury Secretary Janet Yellen on Tuesda...   
22  Workers are seen at the damaged Kharkiv Combin...   
23  A view shows the Zaporizhzhia Nuclear Power Pl...   
24  New Fine Gael leader Simon Harris at Aras an U...   
25  CHARLESTON, W.Va. (AP) — Don Blankenship hasn’...   
26  Kyrylo Budanov, chief of the Main Directorate ...   
27  CNN —\n\nThe Biden administration is preparing...   
28  CNN —\n\nProsecutor Chris Conroy captured the ...   
29  CNN —\n\nThe Republican National Committee las...   

                      Media     Label  
0      https://www.cnbc.com  Politics  
1      https://www.cnbc.com  Politics  
2      https://www.cnbc.com  Politics  
3        https://apnews.com  Politics  
4      https://www.cnbc.com  Politics  
5      https://www.cnbc.com  Politics  
6      https://www.cnbc.com  Politics  
7      https://www.cnbc.com  Politics  
8      https://www.cnbc.com  Politics  
9      https://www.cnbc.com  Politics  
10     https://www.cnbc.com  Politics  
11  https://edition.cnn.com  Politics  
12     https://www.cnbc.com  Politics  
13  https://edition.cnn.com  Politics  
14       https://apnews.com  Politics  
15  https://edition.cnn.com  Politics  
16  https://edition.cnn.com  Politics  
17  https://edition.cnn.com  Politics  
18  https://edition.cnn.com  Politics  
19     https://www.cnbc.com  Politics  
20     https://www.cnbc.com  Politics  
21     https://www.cnbc.com  Politics  
22     https://www.cnbc.com  Politics  
23     https://www.cnbc.com  Politics  
24     https://www.cnbc.com  Politics  
25       https://apnews.com  Politics  
26     https://www.cnbc.com  Politics  
27  https://edition.cnn.com  Politics  
28  https://edition.cnn.com  Politics  
29  https://edition.cnn.com  Politics

## Sports

In [ ]:
sports_links = [
'https://apnews.com/article/chinese-grand-prix-verstappen-f1-d99d6602412f48859ceea704add6a188',
'https://apnews.com/article/capitals-rangers-preview-playoffs-c043bcc15e7f28b03cd9c983a873b569',
'https://edition.cnn.com/2022/02/01/sport/valorant-esports-women-diversity-spt-intl/index.html',
'https://edition.cnn.com/2024/04/08/sport/san-quentin-marathon-intl-spt/index.html',
'https://apnews.com/article/real-madrid-barcelona-xavi-ancelotti-cdac4cc2cd754a45f886ff23c34a2cb0',
'https://apnews.com/article/nashville-miami-score-e2d2b5e250ce6050ae4c32bf0c899929',
'https://apnews.com/article/napoli-serie-a-a1bb1594099fd146c858b7a2ea67ad0d',
'https://apnews.com/article/chinese-gp-formula-1-sprint-race-7b037c184f9ad9b1b02c6540cb726d32',
'https://apnews.com/article/fifa-player-agents-infantino-6d82ee3a9959760e5d477bbd99c54110',
'https://apnews.com/article/womens-champions-league-barcelona-chelsea-psg-lyon-ad2abc08fb3fd690ebca7dac8abbe906',
'https://apnews.com/article/lakers-nuggets-jokic-gru-5b44fdb92642991513194d8e75c40411',
'https://apnews.com/philadelphia-76ers-new-york-knicks-nba-playoffs-new-york-7bc72cfe265a4f9fa1d949100aeaf01e',
'https://apnews.com/denver-nuggets-los-angeles-lakers-nba-playoffs-a051a5db3e2642a5aa116fd5a0b356a2',
'https://apnews.com/article/suns-timberwolves-edwards-durant-911878935cf0b3fc7780cd7a67a2a5bc',
'https://apnews.com/article/suns-timberwolves-score-edwards-460a8f225a95cf9e7c7630e1aabd373c',
'https://apnews.com/article/magic-cavaliers-score-546636bbf9ed5e79addc53582757da3d',
'https://apnews.com/article/lakers-nuggets-nba-playoffs-2024-9463cdf14f3d5696d92f1b7ab3ff3403',
'https://apnews.com/miami-heat-boston-celtics-nba-playoffs-dc800fcf138b4bf0b60c94534138ae2c',
'https://apnews.com/article/los-angeles-clippers-fined-ba597e6ccfaec699156d1e5711ce6706',
'https://apnews.com/article/maulik-pancholy-pennsylvania-school-speech-canceled-e01d4194c88bb7dc9e298f411d804be4',
'https://apnews.com/article/wolves-arsenal-score-8554895f4c9d4a8e6ac5652612090e86',
'https://apnews.com/article/rbc-heritage-collin-morikawa-scottie-scheffler-fe979283508f4a443bda3366cdf1ed63',
'https://apnews.com/article/wesley-bryan-pga-tour-dominican-puntacana-618e658ae95f28483aaf35d8ab11bb66',
'https://apnews.com/article/lpga-korda-chevron-240a67972e871233afea9f7ec13fa5e4',
'https://apnews.com/article/olympics-ai-artificial-intelligence-sports-30d3a33d4be893e1495e63fc1116aecc',
'https://apnews.com/article/asian-champions-league-score-eb5dd138c693527d6bf368d8e63ae1b8',
'https://edition.cnn.com/2024/04/19/sport/jurgen-klopp-mohamed-salah-liverpool-europa-league-spt-intl/index.html',
'https://edition.cnn.com/2024/04/19/sport/bayer-leverkusen-xabi-alonso-2024-spt-intl/index.html',
'https://edition.cnn.com/2024/03/29/sport/three-people-arrested-threatening-angel-di-maria-spt-intl/index.html',
'https://edition.cnn.com/2024/03/19/sport/real-madrid-files-complaint-against-referee-vinicius-jr-spt-intl/index.html'


]


In [ ]:
sports_df = pd.DataFrame()

for link in sports_links:
  article = Article(link)
  article.download()
  article.parse()

  list_text = []
  list_source = []
  list_label = []

  list_text.append(article.text)
  list_source.append(article.source_url)
  list_label.append('Sports')

  temp_df = pd.DataFrame({'Text': list_text, 'Media': list_source, 'Label': list_label})
  sports_df = pd.concat([sports_df, temp_df], ignore_index=True)


In [ ]:
sports_df

Text  \
0   SHANGHAI (AP) — The Chinese Grand Prix could d...   
1   The New York Rangers were the first team to cl...   
2   CNN —\n\nThose within esports often get tired ...   
3   CNN —\n\nIn San Quentin State Prison and decad...   
4   BARCELONA, Spain (AP) — Real Madrid faces one ...   
5   FORT LAUDERDALE, Fla. (AP) — Lionel Messi scor...   
6   ROME (AP) — Napoli’s title defense is going fr...   
7   SHANGHAI (AP) — Max Verstappen, continuing his...   
8   ZURICH (AP) — FIFA has called on lawmakers to ...   
9   BARCELONA, Spain (AP) — Chelsea handed defendi...   
10  DENVER (AP) — Nikola Jokic arrived in characte...   
11  Philadelphia 76ers (47-35, seventh in the East...   
12  Los Angeles Lakers (47-35, eighth in the Weste...   
13  MINNEAPOLIS (AP) — Anthony Edwards drained a 3...   
14  MINNEAPOLIS (AP) — Anthony Edwards scored 18 o...   
15  CLEVELAND (AP) — Donovan Mitchell and the Cava...   
16  A year ago when the Los Angeles Lakers faced t...   
17  Miami Heat (46-36, eighth in the Eastern Confe...   
18  NEW YORK (AP) — The Los Angeles Clippers were ...   
19  MECHANICSBURG, Pa. (AP) — A Pennsylvania schoo...   
20  WOLVERHAMPTON, England (AP) — Arsenal moved ba...   
21  HILTON HEAD ISLAND, S.C. (AP) — Collin Morikaw...   
22  PUNTA CANA, Dominican Republic. (AP) — Wesley ...   
23  THE WOODLANDS, Texas (AP) — Nelly Korda hasn’t...   
24  LONDON (AP) — Olympic organizers unveiled thei...   
25  RIYADH, Saudi Arabia (AP) — Al-Ain of the Unit...   
26  CNN —\n\nLiverpool manager Jürgen Klopp said h...   
27  CNN —\n\n‘Jurassic Park’ has just been release...   
28  CNN —\n\nThree people were arrested in Argenti...   
29  CNN —\n\nReal Madrid says it has filed a compl...   

                      Media   Label  
0        https://apnews.com  Sports  
1        https://apnews.com  Sports  
2   https://edition.cnn.com  Sports  
3   https://edition.cnn.com  Sports  
4        https://apnews.com  Sports  
5        https://apnews.com  Sports  
6        https://apnews.com  Sports  
7        https://apnews.com  Sports  
8        https://apnews.com  Sports  
9        https://apnews.com  Sports  
10       https://apnews.com  Sports  
11       https://apnews.com  Sports  
12       https://apnews.com  Sports  
13       https://apnews.com  Sports  
14       https://apnews.com  Sports  
15       https://apnews.com  Sports  
16       https://apnews.com  Sports  
17       https://apnews.com  Sports  
18       https://apnews.com  Sports  
19       https://apnews.com  Sports  
20       https://apnews.com  Sports  
21       https://apnews.com  Sports  
22       https://apnews.com  Sports  
23       https://apnews.com  Sports  
24       https://apnews.com  Sports  
25       https://apnews.com  Sports  
26  https://edition.cnn.com  Sports  
27  https://edition.cnn.com  Sports  
28  https://edition.cnn.com  Sports  
29  https://edition.cnn.com  Sports

## Entertainment

In [ ]:
entertainment_links = [
'https://www.tmz.com/2024/04/20/taylor-swift-fortnight-music-video-travis-kelce-tortured-poets-department/#continued',
'https://edition.cnn.com/2024/04/19/entertainment/orlando-bloom-katy-perry/index.html',
'https://edition.cnn.com/2024/04/19/entertainment/jenna-fischer-mean-girls/index.html',
'https://edition.cnn.com/2024/04/11/entertainment/civil-war-movie-review/index.html',
'https://www.ign.com/articles/fallout-4-sales-top-ten-europe',
'https://kotaku.com/deathtroopers-fan-game-star-wars-zombies-stormtroopers-1851420306',
'https://www.gamespot.com/articles/save-20-on-no-rest-for-the-wicked-and-get-a-free-game-for-a-limited-time/1100-6522785/',
'https://www.gamespot.com/articles/asus-customizable-xbox-controller-drops-to-lowest-price-yet-at-amazon/1100-6522816/',
'https://www.gamespot.com/gallery/best-reviewed-games-2023/2900-4922/#57',
'https://www.animenewsnetwork.com/interest/2024-04-18/frieren-beyond-journey-end-series-gets-escape-room-set-in-dream-world/.209987',
'https://www.ign.com/articles/how-honkai-star-rail-is-using-ai-technology-to-supplement-development',
'https://www.ign.com/articles/genshin-impact-fans-show-off-your-art-and-win',
'https://edition.cnn.com/2024/04/18/entertainment/ministry-of-ungentlemanly-warfare-review-henry-cavill/index.html',
'https://edition.cnn.com/cnn-underscored/reviews/call-of-duty-modern-warfare-3-2023?iid=cnn_buildContentRecirc_end_recirc',
'https://edition.cnn.com/2024/04/09/entertainment/robert-downey-jr-kimmel-joke/index.html',
'https://www.ign.com/articles/genshin-impact-fans-show-off-your-art-and-win',
'https://apnews.com/article/morgan-wallen-arrest-bar-message-12788b00384ed365ab75e12d5b166dee',
'https://apnews.com/article/jonathan-tetelman-tenor-17333d70486a3207d5d7c311620f5ea8',
'https://www.ign.com/articles/silent-hill-2-remakes-main-character-seemingly-redesigned-and-fans-are-actually-pretty-happy',
'https://www.ign.com/articles/razer-kishi-ultra-ipad-mini-usb-c-support',
'https://kotaku.com/kingdom-come-deliverance-2-size-release-date-1851420482',
'https://kotaku.com/the-witcher-final-season-production-1851419742',
'https://kotaku.com/overwatch-2-queue-times-ranked-wide-group-season-10-1851415862',
'https://kotaku.com/palworld-tencent-pokemon-nintendo-clone-mobile-1851415602',
'https://www.gamespot.com/articles/breaking-bads-anna-gunn-says-attitudes-toward-skyler-have-now-changed/1100-6522806/',
'https://kotaku.com/palworld-pokemon-lawsuit-pocketpair-nintendo-pc-xbox-1851184907',
'https://kotaku.com/palworld-continues-to-grow-as-it-shoots-past-25-million-1851279807',
'https://kotaku.com/palworld-shoot-pokemon-with-guns-impressions-review-pc-1851185681',
'https://edition.cnn.com/2024/04/20/entertainment/phish-concerts-sphere-anastasio-cec/index.html',
'https://edition.cnn.com/2024/04/20/entertainment/the-jinx-part-two-review-robert-durst/index.html'


]


In [ ]:
entertainment_df = pd.DataFrame()

for link in entertainment_links:
  article = Article(link)
  article.download()
  article.parse()

  list_text = []
  list_source = []
  list_label = []

  list_text.append(article.text)
  list_source.append(article.source_url)
  list_label.append('Entertainment')

  temp_df = pd.DataFrame({'Text': list_text, 'Media': list_source, 'Label': list_label})
  entertainment_df = pd.concat([entertainment_df, temp_df], ignore_index=True)


In [ ]:
entertainment_df

Text  \
0   Taylor Swift can't hide she's totally smitten ...   
1   CNN —\n\nOrlando Bloom knows successful partne...   
2   CNN —\n\nJenna Fischer can perform through pai...   
3   CNN —\n\nThe details of “Civil War” don’t make...   
4   Fallout 4, the post-apocalyptic wasteland acti...   
5   A short Star Wars fan game about a stormtroope...   
6   The gorgeous action-RPG No Rest for the Wicked...   
7   The ROG Raikiri and Raikiri Pro Xbox controlle...   
8   GameSpot may receive revenue from affiliate an...   
9   Teams have 1 hour to wake up Frieren\n\nSCRAP ...   
10  More and more game studios are actively discus...   
11  In celebration of Genshin Impact Genesis Cryst...   
12  CNN —\n\nAs war movies go, “The Ministry of Un...   
13  Following the release of last year’s Call of D...   
14  CNN —\n\nRobert Downey Jr. was unfazed by Jimm...   
15  In celebration of Genshin Impact Genesis Cryst...   
16  NASHVILLE, Tenn. (AP) — Country music star Mor...   
17  NEW YORK (AP) — Jonathan Tetelman transformed ...   
18  It looks like publisher Konami and developer B...   
19  Razer continues to support the mobile market, ...   
20  Warhorse Studios announced Kingdom Come: Deliv...   
21  It is official: Netflix’s series based on The ...   
22  Overwatch 2’s tenth season is officially under...   
23  After selling over 15 million copies in its fi...   
24  Anna Gunn famously played Skyler White on Brea...   
25  We’re about done with the first month of the y...   
26  Hey, remember that little game Palworld? Well ...   
27  Over the weekend, I downloaded Palworld on my ...   
28  CNN —\n\nConcerts by Phish, the beloved Vermon...   
29  CNN —\n\nThe first thought hearing about “The ...   

                               Media          Label  
0                https://www.tmz.com  Entertainment  
1            https://edition.cnn.com  Entertainment  
2            https://edition.cnn.com  Entertainment  
3            https://edition.cnn.com  Entertainment  
4                https://www.ign.com  Entertainment  
5                 https://kotaku.com  Entertainment  
6           https://www.gamespot.com  Entertainment  
7           https://www.gamespot.com  Entertainment  
8           https://www.gamespot.com  Entertainment  
9   https://www.animenewsnetwork.com  Entertainment  
10               https://www.ign.com  Entertainment  
11               https://www.ign.com  Entertainment  
12           https://edition.cnn.com  Entertainment  
13           https://edition.cnn.com  Entertainment  
14           https://edition.cnn.com  Entertainment  
15               https://www.ign.com  Entertainment  
16                https://apnews.com  Entertainment  
17                https://apnews.com  Entertainment  
18               https://www.ign.com  Entertainment  
19               https://www.ign.com  Entertainment  
20                https://kotaku.com  Entertainment  
21                https://kotaku.com  Entertainment  
22                https://kotaku.com  Entertainment  
23                https://kotaku.com  Entertainment  
24          https://www.gamespot.com  Entertainment  
25                https://kotaku.com  Entertainment  
26                https://kotaku.com  Entertainment  
27                https://kotaku.com  Entertainment  
28           https://edition.cnn.com  Entertainment  
29           https://edition.cnn.com  Entertainment

In [ ]:
entertainment_df.to_csv("entertainment.csv")

## Final Dataframe

Setelah melakukan scraping untuk semua berita yang ada dan sudah tersimpan ke dalam dataframe, tiap dataframe tersebut akan digabung menjadi satu kesatuan dataframe terakhir dengan nama **final_df**.

In [ ]:
final_df = pd.DataFrame()
final_df = pd.concat([business_df, politics_df, sports_df, entertainment_df], ignore_index=True)

In [ ]:
final_df

Text  \
0    New York CNN —\n\nHyundai said on Thursday it ...   
1    NEW YORK (AP) — The worst week for big technol...   
2    A couple sits in front of a television with th...   
3    Autonomous delivery drone startup Zipline said...   
4    Illuminated trademark of the American athletic...   
..                                                 ...   
115  We’re about done with the first month of the y...   
116  Hey, remember that little game Palworld? Well ...   
117  Over the weekend, I downloaded Palworld on my ...   
118  CNN —\n\nConcerts by Phish, the beloved Vermon...   
119  CNN —\n\nThe first thought hearing about “The ...   

                       Media          Label  
0    https://edition.cnn.com       Business  
1         https://apnews.com       Business  
2       https://www.cnbc.com       Business  
3       https://www.cnbc.com       Business  
4       https://www.cnbc.com       Business  
..                       ...            ...  
115       https://kotaku.com  Entertainment  
116       https://kotaku.com  Entertainment  
117       https://kotaku.com  Entertainment  
118  https://edition.cnn.com  Entertainment  
119  https://edition.cnn.com  Entertainment  

[120 rows x 3 columns]

Setelah membuat dataframe gabungan yang terakhir, dengan total 120 baris data, maka dataframe tersebut diubah menjadi file .csv

In [ ]:
final_df.to_csv('scraped_data.csv')

# 2. Cleaning

## Cleaning Data

Data yang tadi sudah ada di dalam dataframe akan memasuki proses pembersihan atau cleaning, dimana kita akan mengubah data menjadi huruf kecil semua, mengubah angka menjadi kosong, mengubah karakter yang bukan tergolong huruf menjadi kosong, mengganti whitespace menjadi spasi atau kosong, dan mengganti underscore menjadi spasi kosong. Hasil dari function ini nanti akan dimasukkan menjadi kolom baru untuk dataframe final_df.

In [ ]:
## Function Cleaning Dasar
def cleaning(dataframe):
    df_clean = dataframe.str.lower()
    df_clean = [re.sub(r"\d+","",i )for i in df_clean]
    df_clean = [re.sub(r'[^\w]', ' ', i)for i in df_clean]
    df_clean = [re.sub(r'_',' ',i)for i in df_clean]
    df_clean = [re.sub(r'\s+',' ',i)for i in df_clean]

    return df_clean

In [ ]:
clean_text = cleaning(final_df['Text'])
clean_text

['new york cnn hyundai said on thursday it paused advertising on x formerly known as twitter a day after an ad from the automaker reportedly appeared adjacent to an antisemitic and pro hitler post on the social media site on wednesday user nancy levine stearns posted screenshots of a paid hyundai ad above an antisemitic post from a verified premium account the account posts tweets that deny the holocaust and perpetuate antisemitic rhetoric in response to a cnn inquiry that included stearns post a spokesperson for hyundai replied we have paused our ads on x and are speaking to x directly about brand safety to ensure this issue is addressed hyundai is the latest prominent company to pause advertising on x after its owner elon musk publicly embraced an antisemitic theory favored by white supremacists in nbc news first reported the pause x has suspended the antisemitic account that was adjacent to the ad the head of business operations at x joe benarroch said in an email to cnn the account

In [ ]:
final_df.insert(1, 'Clean Text', clean_text)

In [ ]:
final_df

Text  \
0    New York CNN —\n\nHyundai said on Thursday it ...   
1    NEW YORK (AP) — The worst week for big technol...   
2    A couple sits in front of a television with th...   
3    Autonomous delivery drone startup Zipline said...   
4    Illuminated trademark of the American athletic...   
..                                                 ...   
115  We’re about done with the first month of the y...   
116  Hey, remember that little game Palworld? Well ...   
117  Over the weekend, I downloaded Palworld on my ...   
118  CNN —\n\nConcerts by Phish, the beloved Vermon...   
119  CNN —\n\nThe first thought hearing about “The ...   

                                            Clean Text  \
0    new york cnn hyundai said on thursday it pause...   
1    new york ap the worst week for big technology ...   
2    a couple sits in front of a television with th...   
3    autonomous delivery drone startup zipline said...   
4    illuminated trademark of the american athletic...   
..                                                 ...   
115  we re about done with the first month of the y...   
116  hey remember that little game palworld well as...   
117  over the weekend i downloaded palworld on my p...   
118  cnn concerts by phish the beloved vermont jam ...   
119  cnn the first thought hearing about the jinx p...   

                       Media          Label  
0    https://edition.cnn.com       Business  
1         https://apnews.com       Business  
2       https://www.cnbc.com       Business  
3       https://www.cnbc.com       Business  
4       https://www.cnbc.com       Business  
..                       ...            ...  
115       https://kotaku.com  Entertainment  
116       https://kotaku.com  Entertainment  
117       https://kotaku.com  Entertainment  
118  https://edition.cnn.com  Entertainment  
119  https://edition.cnn.com  Entertainment  

[120 rows x 4 columns]

## Splitting

Setelah cleaning, dataframe final_df akan dipecah menjadi data train dan data testing dengan perbandingan 8:2, dengan train mengambil 80% data dan test mengambil 20% data. Saya juga menggunakan random_state dengan nilai 41 agar terpecah acak.

In [ ]:
## Train Test Split Dataframe
train_df, test_df = train_test_split(final_df, test_size = 0.2, random_state = 41)

## Tokenization

Tokenization adalah proses memecah kalimat yang ada menjadi pecahan kecil berupa kata-kata, frasa, atau sebuah elemen yang bisa digunakan sebagai token. Token ini adalah representasi satu buah informasi.

Ini perlu dilakukan karena pemecahan kalimat menjadi kata-kata mempermudah metode-metode selanjutnya, seperti filtering atau stopwords removal, dan classification. Salah satu cara yang bisa dilakukan untuk tokenization adalah dengan menggunakan function **word_tokenize** dari library **nltk.tokenize**

### Train Tokenization

In [ ]:
## Define a function for tokenization
def tokenizing(text):
  return word_tokenize(text)

train_df['tokenized'] = train_df['Clean Text'].apply(tokenizing)

In [ ]:
train_df

Text  \
113  After selling over 15 million copies in its fi...   
0    New York CNN —\n\nHyundai said on Thursday it ...   
38   FBI Director Christopher Wray testifies before...   
78   NEW YORK (AP) — The Los Angeles Clippers were ...   
37   Ukrainian President Volodymyr Zelenskyy attend...   
..                                                 ...   
80   WOLVERHAMPTON, England (AP) — Arsenal moved ba...   
98   GameSpot may receive revenue from affiliate an...   
12   NEW YORK (AP) — For many people, tax season is...   
35   A majority of Southeast Asians would align wit...   
64   BARCELONA, Spain (AP) — Real Madrid faces one ...   

                                            Clean Text  \
113  after selling over million copies in its first...   
0    new york cnn hyundai said on thursday it pause...   
38   fbi director christopher wray testifies before...   
78   new york ap the los angeles clippers were fine...   
37   ukrainian president volodymyr zelenskyy attend...   
..                                                 ...   
80   wolverhampton england ap arsenal moved back to...   
98   gamespot may receive revenue from affiliate an...   
12   new york ap for many people tax season isn t o...   
35   a majority of southeast asians would align wit...   
64   barcelona spain ap real madrid faces one last ...   

                        Media          Label  \
113        https://kotaku.com  Entertainment   
0     https://edition.cnn.com       Business   
38       https://www.cnbc.com       Politics   
78         https://apnews.com         Sports   
37       https://www.cnbc.com       Politics   
..                        ...            ...   
80         https://apnews.com         Sports   
98   https://www.gamespot.com  Entertainment   
12         https://apnews.com       Business   
35       https://www.cnbc.com       Politics   
64         https://apnews.com         Sports   

                                             tokenized  
113  [after, selling, over, million, copies, in, it...  
0    [new, york, cnn, hyundai, said, on, thursday, ...  
38   [fbi, director, christopher, wray, testifies, ...  
78   [new, york, ap, the, los, angeles, clippers, w...  
37   [ukrainian, president, volodymyr, zelenskyy, a...  
..                                                 ...  
80   [wolverhampton, england, ap, arsenal, moved, b...  
98   [gamespot, may, receive, revenue, from, affili...  
12   [new, york, ap, for, many, people, tax, season...  
35   [a, majority, of, southeast, asians, would, al...  
64   [barcelona, spain, ap, real, madrid, faces, on...  

[96 rows x 5 columns]

### Test Tokenization

In [ ]:
test_df['tokenized'] = test_df['Clean Text'].apply(tokenizing)

## Filtering

Filtering atau proses stopword removal adalah tahap text preprocessing yang digunakan untuk menghilangkan "stopwords" atau kata-kata yang umum digunakan sampai ditandai sebagai tidak berisi informasi penting. Kata-kata ini biasanya dianggap hanya sebagai pelengkap atau pendukung untuk kata yang memang mempunyai informasi penting. Misalnya di bahasa Inggris, sering ditemukan kata-kata seperti "the", "is", "in", atau yang lainnya.

Stopwords removal dilakukan karena membantu dalam pengurangan noise atau menghilangkan data tidak penting agar lebih mudah kita menemukan informasi penting yang dibutuhkan. Selain itu, stopwords removal juga mengurangi dimensi pada fitur kita, sehingga membuat model bisa fokus kepada informasi atau kata-kata penting.

### Train Filtering

In [ ]:
## Filtering done by removing stopwords

from nltk.corpus import stopwords
nltk.download('stopwords')

stopwords_list = set(stopwords.words('english'))

## Function to Remove Stopwords
def stopword_removal(tokenized):
  return [word for word in tokenized if word.lower() not in stopwords_list]

train_df['filtered'] = train_df['tokenized'].apply(stopword_removal)

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


In [ ]:
train_df.drop(columns=['Media'], inplace = True)

### Test Filtering

In [ ]:
test_df['filtered'] = test_df['tokenized'].apply(stopword_removal)

## Stemming or Lemmatization

Stemming and Lemmatization adalah teknik normalisasi kata-kata di text mining untuk mengubah suatu kata menjadi bentuk dasar mereka. Stemming dan Lemmatization mempunyai tujuan sama, tetapi hanya berbeda pada metode perubahan mereka. Stemming hanya mengubahnya dengan memotong imbuhan tambahan dari kata, sedangkan Lemmatization mengubahnya ke bentuk dasar dengan mempertimbangkan bentuk pada kamus kata yang digunakan.

Untuk kasus seperti ini, saya menggunakan Lemmatization karena data berasal dari berita-berita dimana konteks setiap katanya penting tergantung pemakaiannya dan dimana dipakainya. Tujuan penggunaan metode ini adalah untuk normalisasi yang membantu mengurangi jumlah fitur agar model bisa bekerja lebih efisien.

### Train Lemmatization

In [ ]:
## Use Lemmatization because we want a more accurate transformation of the words
nltk.download('wordnet')
from nltk import WordNetLemmatizer
lemma = WordNetLemmatizer()

def lemmatizer(tokens):
  return[lemma.lemmatize(word) for word in tokens]

train_df['lemmatized'] = train_df['filtered'].apply(lemmatizer)

[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


In [ ]:
train_df

Text  \
113  After selling over 15 million copies in its fi...   
0    New York CNN —\n\nHyundai said on Thursday it ...   
38   FBI Director Christopher Wray testifies before...   
78   NEW YORK (AP) — The Los Angeles Clippers were ...   
37   Ukrainian President Volodymyr Zelenskyy attend...   
..                                                 ...   
80   WOLVERHAMPTON, England (AP) — Arsenal moved ba...   
98   GameSpot may receive revenue from affiliate an...   
12   NEW YORK (AP) — For many people, tax season is...   
35   A majority of Southeast Asians would align wit...   
64   BARCELONA, Spain (AP) — Real Madrid faces one ...   

                                            Clean Text          Label  \
113  after selling over million copies in its first...  Entertainment   
0    new york cnn hyundai said on thursday it pause...       Business   
38   fbi director christopher wray testifies before...       Politics   
78   new york ap the los angeles clippers were fine...         Sports   
37   ukrainian president volodymyr zelenskyy attend...       Politics   
..                                                 ...            ...   
80   wolverhampton england ap arsenal moved back to...         Sports   
98   gamespot may receive revenue from affiliate an...  Entertainment   
12   new york ap for many people tax season isn t o...       Business   
35   a majority of southeast asians would align wit...       Politics   
64   barcelona spain ap real madrid faces one last ...         Sports   

                                             tokenized  \
113  [after, selling, over, million, copies, in, it...   
0    [new, york, cnn, hyundai, said, on, thursday, ...   
38   [fbi, director, christopher, wray, testifies, ...   
78   [new, york, ap, the, los, angeles, clippers, w...   
37   [ukrainian, president, volodymyr, zelenskyy, a...   
..                                                 ...   
80   [wolverhampton, england, ap, arsenal, moved, b...   
98   [gamespot, may, receive, revenue, from, affili...   
12   [new, york, ap, for, many, people, tax, season...   
35   [a, majority, of, southeast, asians, would, al...   
64   [barcelona, spain, ap, real, madrid, faces, on...   

                                              filtered  \
113  [selling, million, copies, first, month, mania...   
0    [new, york, cnn, hyundai, said, thursday, paus...   
38   [fbi, director, christopher, wray, testifies, ...   
78   [new, york, ap, los, angeles, clippers, fined,...   
37   [ukrainian, president, volodymyr, zelenskyy, a...   
..                                                 ...   
80   [wolverhampton, england, ap, arsenal, moved, b...   
98   [gamespot, may, receive, revenue, affiliate, a...   
12   [new, york, ap, many, people, tax, season, gat...   
35   [majority, southeast, asians, would, align, ch...   
64   [barcelona, spain, ap, real, madrid, faces, on...   

                                            lemmatized  
113  [selling, million, copy, first, month, mania, ...  
0    [new, york, cnn, hyundai, said, thursday, paus...  
38   [fbi, director, christopher, wray, testifies, ...  
78   [new, york, ap, los, angeles, clipper, fined, ...  
37   [ukrainian, president, volodymyr, zelenskyy, a...  
..                                                 ...  
80   [wolverhampton, england, ap, arsenal, moved, b...  
98   [gamespot, may, receive, revenue, affiliate, a...  
12   [new, york, ap, many, people, tax, season, gat...  
35   [majority, southeast, asian, would, align, chi...  
64   [barcelona, spain, ap, real, madrid, face, one...  

[96 rows x 6 columns]

### Test Lemmatization

In [ ]:
test_df['lemmatized'] = test_df['filtered'].apply(lemmatizer)

# 3. Text Representation

Tahap ini adalah pengubahan kata-kata dari data yang sudah di token menjadi angka-angka. Hal ini dilakukan dengan tujuan agar kata-kata tersebut dapat diproses oleh model machine learning.

## TF-IDF

TF-IDF adalah teknik dimana kata-kata diberikan sebuah weight berdasarkan frekuensi atau keseringan munculnya kata tersebut dalam data yang nantinya dibandingkan dengan seluruh korpus dokumen atau semua dokumen. Semakin besar frekuensi kata tersebut, maka weightnya semakin kecil, menandakan semakin tidak penting atau informatif.


Sumber:

https://towardsdatascience.com/how-tf-idf-works-3dbf35e568f0

In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer

In [ ]:
vector = TfidfVectorizer()
tfidf_train = vector.fit_transform(train_df["Clean Text"])
TFIDF_train = pd.DataFrame(tfidf_train.toarray(), columns=vector.get_feature_names_out())
TFIDF_train

aaa  aaron  aba  abandon  abandoned  abc  abe  aberg  abigail  \
0   0.000000    0.0  0.0      0.0        0.0  0.0  0.0    0.0      0.0   
1   0.000000    0.0  0.0      0.0        0.0  0.0  0.0    0.0      0.0   
2   0.000000    0.0  0.0      0.0        0.0  0.0  0.0    0.0      0.0   
3   0.000000    0.0  0.0      0.0        0.0  0.0  0.0    0.0      0.0   
4   0.000000    0.0  0.0      0.0        0.0  0.0  0.0    0.0      0.0   
..       ...    ...  ...      ...        ...  ...  ...    ...      ...   
91  0.000000    0.0  0.0      0.0        0.0  0.0  0.0    0.0      0.0   
92  0.087105    0.0  0.0      0.0        0.0  0.0  0.0    0.0      0.0   
93  0.000000    0.0  0.0      0.0        0.0  0.0  0.0    0.0      0.0   
94  0.000000    0.0  0.0      0.0        0.0  0.0  0.0    0.0      0.0   
95  0.000000    0.0  0.0      0.0        0.0  0.0  0.0    0.0      0.0   

     ability  ...  zone  zootopia  zurich  zwilling  ángel  アベツカサ  小学館  山田鐘人  \
0   0.000000  ...   0.0       0.0     0.0       0.0    0.0    0.0  0.0   0.0   
1   0.000000  ...   0.0       0.0     0.0       0.0    0.0    0.0  0.0   0.0   
2   0.036889  ...   0.0       0.0     0.0       0.0    0.0    0.0  0.0   0.0   
3   0.000000  ...   0.0       0.0     0.0       0.0    0.0    0.0  0.0   0.0   
4   0.000000  ...   0.0       0.0     0.0       0.0    0.0    0.0  0.0   0.0   
..       ...  ...   ...       ...     ...       ...    ...    ...  ...   ...   
91  0.000000  ...   0.0       0.0     0.0       0.0    0.0    0.0  0.0   0.0   
92  0.000000  ...   0.0       0.0     0.0       0.0    0.0    0.0  0.0   0.0   
93  0.000000  ...   0.0       0.0     0.0       0.0    0.0    0.0  0.0   0.0   
94  0.000000  ...   0.0       0.0     0.0       0.0    0.0    0.0  0.0   0.0   
95  0.000000  ...   0.0       0.0     0.0       0.0    0.0    0.0  0.0   0.0   

    葬送のフリーレン  製作委員会  
0        0.0    0.0  
1        0.0    0.0  
2        0.0    0.0  
3        0.0    0.0  
4        0.0    0.0  
..       ...    ...  
91       0.0    0.0  
92       0.0    0.0  
93       0.0    0.0  
94       0.0    0.0  
95       0.0    0.0  

[96 rows x 7958 columns]

In [ ]:
## Setting Up the Test Data for TFIDF
test_tfidf = vector.transform(test_df["Clean Text"])
TFIDF_test = pd.DataFrame(test_tfidf.toarray(), columns = vector.get_feature_names_out())
TFIDF_test

aaa  aaron  aba  abandon  abandoned  abc  abe  aberg  abigail  \
0   0.000000    0.0  0.0      0.0        0.0  0.0  0.0    0.0      0.0   
1   0.000000    0.0  0.0      0.0        0.0  0.0  0.0    0.0      0.0   
2   0.000000    0.0  0.0      0.0        0.0  0.0  0.0    0.0      0.0   
3   0.000000    0.0  0.0      0.0        0.0  0.0  0.0    0.0      0.0   
4   0.000000    0.0  0.0      0.0        0.0  0.0  0.0    0.0      0.0   
5   0.000000    0.0  0.0      0.0        0.0  0.0  0.0    0.0      0.0   
6   0.000000    0.0  0.0      0.0        0.0  0.0  0.0    0.0      0.0   
7   0.000000    0.0  0.0      0.0        0.0  0.0  0.0    0.0      0.0   
8   0.000000    0.0  0.0      0.0        0.0  0.0  0.0    0.0      0.0   
9   0.000000    0.0  0.0      0.0        0.0  0.0  0.0    0.0      0.0   
10  0.000000    0.0  0.0      0.0        0.0  0.0  0.0    0.0      0.0   
11  0.000000    0.0  0.0      0.0        0.0  0.0  0.0    0.0      0.0   
12  0.000000    0.0  0.0      0.0        0.0  0.0  0.0    0.0      0.0   
13  0.000000    0.0  0.0      0.0        0.0  0.0  0.0    0.0      0.0   
14  0.000000    0.0  0.0      0.0        0.0  0.0  0.0    0.0      0.0   
15  0.000000    0.0  0.0      0.0        0.0  0.0  0.0    0.0      0.0   
16  0.000000    0.0  0.0      0.0        0.0  0.0  0.0    0.0      0.0   
17  0.000000    0.0  0.0      0.0        0.0  0.0  0.0    0.0      0.0   
18  0.000000    0.0  0.0      0.0        0.0  0.0  0.0    0.0      0.0   
19  0.000000    0.0  0.0      0.0        0.0  0.0  0.0    0.0      0.0   
20  0.000000    0.0  0.0      0.0        0.0  0.0  0.0    0.0      0.0   
21  0.038213    0.0  0.0      0.0        0.0  0.0  0.0    0.0      0.0   
22  0.000000    0.0  0.0      0.0        0.0  0.0  0.0    0.0      0.0   
23  0.000000    0.0  0.0      0.0        0.0  0.0  0.0    0.0      0.0   

    ability  ...  zone  zootopia  zurich  zwilling  ángel  アベツカサ  小学館  山田鐘人  \
0       0.0  ...   0.0       0.0     0.0       0.0    0.0    0.0  0.0   0.0   
1       0.0  ...   0.0       0.0     0.0       0.0    0.0    0.0  0.0   0.0   
2       0.0  ...   0.0       0.0     0.0       0.0    0.0    0.0  0.0   0.0   
3       0.0  ...   0.0       0.0     0.0       0.0    0.0    0.0  0.0   0.0   
4       0.0  ...   0.0       0.0     0.0       0.0    0.0    0.0  0.0   0.0   
5       0.0  ...   0.0       0.0     0.0       0.0    0.0    0.0  0.0   0.0   
6       0.0  ...   0.0       0.0     0.0       0.0    0.0    0.0  0.0   0.0   
7       0.0  ...   0.0       0.0     0.0       0.0    0.0    0.0  0.0   0.0   
8       0.0  ...   0.0       0.0     0.0       0.0    0.0    0.0  0.0   0.0   
9       0.0  ...   0.0       0.0     0.0       0.0    0.0    0.0  0.0   0.0   
10      0.0  ...   0.0       0.0     0.0       0.0    0.0    0.0  0.0   0.0   
11      0.0  ...   0.0       0.0     0.0       0.0    0.0    0.0  0.0   0.0   
12      0.0  ...   0.0       0.0     0.0       0.0    0.0    0.0  0.0   0.0   
13      0.0  ...   0.0       0.0     0.0       0.0    0.0    0.0  0.0   0.0   
14      0.0  ...   0.0       0.0     0.0       0.0    0.0    0.0  0.0   0.0   
15      0.0  ...   0.0       0.0     0.0       0.0    0.0    0.0  0.0   0.0   
16      0.0  ...   0.0       0.0     0.0       0.0    0.0    0.0  0.0   0.0   
17      0.0  ...   0.0       0.0     0.0       0.0    0.0    0.0  0.0   0.0   
18      0.0  ...   0.0       0.0     0.0       0.0    0.0    0.0  0.0   0.0   
19      0.0  ...   0.0       0.0     0.0       0.0    0.0    0.0  0.0   0.0   
20      0.0  ...   0.0       0.0     0.0       0.0    0.0    0.0  0.0   0.0   
21      0.0  ...   0.0       0.0     0.0       0.0    0.0    0.0  0.0   0.0   
22      0.0  ...   0.0       0.0     0.0       0.0    0.0    0.0  0.0   0.0   
23      0.0  ...   0.0       0.0     0.0       0.0    0.0    0.0  0.0   0.0   

    葬送のフリーレン  製作委員会  
0        0.0    0.0  
1        0.0    0.0  
2        0.0    0.0  
3        0.0    0.0  
4        0.0    0.0  
5        0.0    0.0  
6        0.0    0.0  
7     

## Word2Vec - Skipgram

**Word2Vec** adalah salah satu metode penyisipan atau penyematan kata yang sering ditemukan di text mining. Cara kerja dari Word2Vec adalah kata-kata sudah digambarkan menjadi satu buah vektor dan vektor ini nantinya akan dipelajari secara neural network. Vektor akan berusaha untuk menangkap karakteristik dari kata dan akan dilihat hubungannya berdasarkan arti, konteks penggunaan, dan lainnya. Hasil ini akan direpresentasikan dalam bentuk angka yang nantinya bisa diproses oleh model.

Sumber:

https://towardsdatascience.com/word2vec-explained-49c52b4ccb71


In [ ]:
import gensim
from gensim.models import Word2Vec

Word2Vec mempunyai dua arsitektur utama, yaitu Continuous Bag of Words (CBOW) dan Skipgram. Untuk kasus ini, saya menggunakan Skipgram, karena tujuan dari skipgram adalah memprediksi context sebuah kata atau data berdasarkan input dalam bentuk kata. Hal ini sangat sesuai dengan kasus ini, dimana kita mempunyai input atau X dalam bentuk kata-kata dari artikel berita dimana kita akan mencoba menggunakan kata-kata tersebut untuk menentukan pengelompokkan berita tersebut ke dalam labelnya.

Model Word2Vec bisa dibuat dengan bantuan library **gensim**. Pembuatan model akan didasarkan pada data training, dimana hasil model akan digunakan untuk membuat sebuah korpus atau kamus yang dipakai untuk memproses data test nantinya.

In [ ]:
## Create Skipgram Model
skipgram_model = gensim.models.Word2Vec(train_df['tokenized'], min_count = 3, vector_size = 50, window = 5, sg=1)

In [ ]:
## Making the vocabulary for skipgram from train data
vocabulary_skipgram = skipgram_model.wv.index_to_key
print(vocabulary_skipgram)

['the', 'to', 'and', 'in', 'of', 'a', 's', 'that', 'on', 'for', 'it', 'is', 'with', 'said', 'as', 'was', 'from', 'its', 'has', 'he', 'at', 'we', 'by', 'i', 'this', 'have', 'his', 'be', 'but', 'an', 'are', 'will', 'year', 'they', 'not', 'after', 't', 'their', 'more', 'one', 'game', 'new', 'about', 'last', 'first', 'you', 'or', 'had', 'up', 'also', 'who', 'trump', 'can', 'company', 'which', 'been', 'over', 'u', 'all', 'would', 'out', 'she', 'when', 'than', 'were', 'time', 'so', 'our', 'like', 'million', 'if', 'what', 'two', 'just', 'ukraine', 'into', 'could', 'while', 'some', 'other', 'games', 'there', 'most', 'season', 'president', 'now', 'quarter', 'billion', 'years', 're', 'through', 'people', 'no', 'china', 'points', 'since', 'only', 'reported', 'ap', 'her', 'cnn', 'second', 'march', 'off', 'because', 'three', 'team', 'how', 'back', 'told', 'where', 'do', 'league', 'growth', 'day', 'us', 'week', 'against', 'still', 'state', 'york', 'going', 'fans', 'win', 'think', 'even', 'much', 'pl

In [ ]:
## Mean vectors
def calculate_mean_vectors(tokens):
  token_vectors = [skipgram_model.wv[token] for token in tokens if token in vocabulary_skipgram]
  if token_vectors:
    return np.mean(token_vectors, axis = 0)
  else:
    return [0]*skipgram_model.vector_size

## Skipgram data frames
skipgram_train = pd.DataFrame()
skipgram_test = pd.DataFrame()

skipgram_train['word_vectors'] = train_df['tokenized'].apply(calculate_mean_vectors)
skipgram_test['word_vectors'] = test_df['tokenized'].apply(calculate_mean_vectors)

columns = [f'{i+1}' for i in range(skipgram_model.vector_size)]

skipgram_train[columns] = pd.DataFrame(skipgram_train['word_vectors'].tolist(), index=skipgram_train.index)
skipgram_test[columns] = pd.DataFrame(skipgram_test['word_vectors'].tolist(), index=skipgram_test.index)

skipgram_train.drop(columns=['word_vectors'], inplace=True)
skipgram_test.drop(columns=['word_vectors'], inplace=True)

In [ ]:
skipgram_train

1         2         3         4         5         6         7  \
113 -0.204160  0.182451 -0.031703  0.045794  0.002120 -0.098509  0.217139   
0   -0.206233  0.190969 -0.019959  0.042117 -0.006997 -0.083884  0.221898   
38  -0.209652  0.185688 -0.030122  0.043441 -0.001570 -0.088227  0.223018   
78  -0.210820  0.208325  0.013588  0.039056 -0.026335 -0.072105  0.239215   
37  -0.215739  0.193801 -0.007227  0.040328 -0.007316 -0.081154  0.236241   
..        ...       ...       ...       ...       ...       ...       ...   
80  -0.216182  0.203390  0.004846  0.039576 -0.004419 -0.084235  0.239895   
98  -0.209439  0.180563 -0.035200  0.046890  0.008931 -0.099905  0.219984   
12  -0.193560  0.169548 -0.057845  0.055641  0.002059 -0.104338  0.205713   
35  -0.216774  0.195258 -0.009027  0.048745  0.004616 -0.086353  0.231330   
64  -0.210472  0.189260 -0.025272  0.043588 -0.002446 -0.097115  0.231365   

            8         9        10  ...        41        42        43  \
113  0.563486 -0.113889 -0.248906  ...  0.201212 -0.092506 -0.256231   
0    0.569898 -0.116062 -0.230222  ...  0.165686 -0.089429 -0.230757   
38   0.564660 -0.116602 -0.227434  ...  0.165124 -0.084482 -0.232697   
78   0.591456 -0.092235 -0.220461  ...  0.147223 -0.066008 -0.204812   
37   0.574964 -0.114532 -0.223339  ...  0.132651 -0.081122 -0.221124   
..        ...       ...       ...  ...       ...       ...       ...   
80   0.591744 -0.108468 -0.230421  ...  0.141584 -0.075245 -0.222410   
98   0.560780 -0.114701 -0.249945  ...  0.197440 -0.090948 -0.254419   
12   0.544297 -0.119433 -0.251384  ...  0.235497 -0.098284 -0.268785   
35   0.578629 -0.113448 -0.228314  ...  0.147025 -0.079823 -0.225709   
64   0.579259 -0.117727 -0.247396  ...  0.187295 -0.088532 -0.253235   

           44        45        46        47        48        49        50  
113 -0.145026  0.247334  0.175508 -0.011811 -0.195301  0.243901  0.266091  
0   -0.158688  0.254795  0.177927  0.002541 -0.187749  0.232420  0.275429  
38  -0.164979  0.251477  0.179878  0.003294 -0.191212  0.239782  0.277921  
78  -0.160618  0.267863  0.222583  0.017360 -0.216274  0.227102  0.321210  
37  -0.163560  0.256583  0.182684  0.016352 -0.190966  0.230701  0.292320  
..        ...       ...       ...       ...       ...       ...       ...  
80  -0.162303  0.261544  0.193830  0.009478 -0.189056  0.221889  0.303632  
98  -0.145737  0.245780  0.175818 -0.010691 -0.193359  0.245510  0.267696  
12  -0.138175  0.250508  0.172460 -0.018257 -0.213562  0.261119  0.260794  
35  -0.166944  0.249556  0.180733  0.009179 -0.183608  0.233349  0.291653  
64  -0.154460  0.254317  0.176946 -0.008006 -0.189494  0.242210  0.282138  

[96 rows x 50 columns]

In [ ]:
skipgram_test

1         2         3         4         5         6         7  \
47  -0.215918  0.191951 -0.025805  0.038757 -0.004429 -0.095233  0.233948   
41  -0.224029  0.199175 -0.012216  0.043299  0.001100 -0.089336  0.238325   
119 -0.212623  0.191458 -0.021862  0.043090  0.000972 -0.097134  0.230612   
43  -0.212694  0.192552 -0.017987  0.040859 -0.006362 -0.090394  0.231614   
33  -0.212062  0.188924 -0.021799  0.044066 -0.004130 -0.091391  0.230359   
73  -0.214005  0.194202 -0.019311  0.040880 -0.001053 -0.101153  0.232738   
16  -0.220308  0.198747 -0.007056  0.047079  0.006006 -0.087667  0.237491   
90  -0.206537  0.189559 -0.017282  0.045642 -0.009908 -0.087547  0.236059   
79  -0.204521  0.182968 -0.036404  0.040409 -0.000885 -0.104153  0.218502   
106 -0.210628  0.190798 -0.020983  0.038470 -0.003351 -0.095291  0.227806   
72  -0.212007  0.211927  0.040467  0.041209 -0.009522 -0.072013  0.239056   
6   -0.223815  0.199783 -0.004559  0.048678  0.002812 -0.085424  0.242369   
48  -0.215031  0.190672 -0.027900  0.039030 -0.004470 -0.096210  0.230898   
62  -0.202878  0.176896 -0.043536  0.049624  0.003021 -0.105491  0.215995   
25  -0.210342  0.190666 -0.010841  0.043677 -0.009311 -0.079526  0.231467   
82  -0.213722  0.197358 -0.008459  0.039524 -0.006362 -0.093055  0.232436   
39  -0.214497  0.194589 -0.017862  0.041346 -0.004896 -0.085835  0.235833   
103 -0.212813  0.189810 -0.019051  0.045577  0.000302 -0.090991  0.229084   
54  -0.217190  0.197608 -0.007848  0.043554 -0.004148 -0.084239  0.237927   
49  -0.218731  0.197826 -0.013244  0.045546 -0.001802 -0.089660  0.235886   
112 -0.204010  0.179961 -0.033398  0.047508  0.003564 -0.102774  0.220894   
27  -0.213500  0.194869 -0.006631  0.045362  0.000512 -0.084052  0.231690   
32  -0.221090  0.197766 -0.006679  0.044425 -0.002085 -0.081863  0.236409   
63  -0.208887  0.183783 -0.036496  0.045513  0.004186 -0.101766  0.223776   

            8         9        10  ...        41        42        43  \
47   0.579650 -0.126055 -0.238320  ...  0.159232 -0.090915 -0.238944   
41   0.583960 -0.119489 -0.233084  ...  0.147561 -0.084837 -0.234847   
119  0.572131 -0.120889 -0.243442  ...  0.169946 -0.088546 -0.244798   
43   0.577708 -0.121021 -0.232568  ...  0.151995 -0.090995 -0.231677   
33   0.577706 -0.120721 -0.238748  ...  0.165597 -0.088347 -0.241255   
73   0.584626 -0.122937 -0.250589  ...  0.175190 -0.089574 -0.254225   
16   0.585973 -0.114221 -0.234604  ...  0.156272 -0.085176 -0.236571   
90   0.577566 -0.117903 -0.235951  ...  0.161274 -0.090159 -0.234458   
79   0.563563 -0.125460 -0.246512  ...  0.188422 -0.092520 -0.255310   
106  0.577394 -0.123571 -0.241270  ...  0.162582 -0.088917 -0.246512   
72   0.611344 -0.096394 -0.235555  ...  0.165624 -0.090637 -0.221002   
6    0.588877 -0.115244 -0.233058  ...  0.139211 -0.081746 -0.226552   
48   0.570436 -0.120977 -0.232629  ...  0.159106 -0.090319 -0.232594   
62   0.561027 -0.120221 -0.256312  ...  0.219075 -0.098188 -0.270205   
25   0.575605 -0.118712 -0.225504  ...  0.150515 -0.084180 -0.231088   
82   0.585270 -0.116387 -0.242200  ...  0.157786 -0.085504 -0.238203   
39   0.580861 -0.119290 -0.228039  ...  0.154849 -0.084305 -0.227451   
103  0.570379 -0.117526 -0.241406  ...  0.172549 -0.088851 -0.245141   
54   0.582054 -0.117307 -0.228685  ...  0.143302 -0.083249 -0.227436   
49   0.587873 -0.117613 -0.236184  ...  0.153116 -0.085690 -0.232091   
112  0.569080 -0.117763 -0.254135  ...  0.204702 -0.094615 -0.264010   
27   0.585605 -0.114711 -0.237616  ...  0.155894 -0.083101 -0.233476   
32   0.588317 -0.115524 -0.224375  ...  0.139614 -0.080911 -0.224453   
63   0.568945 -0.123777 -0.248769  ...  0.191689 -0.094354 -0.259507   

           44        45        46        47        48        49        50  
47  -0.159198  0.261643  0.172189 -0.005527 -0.186291  0.239376  0.285359  
41  -0.162482  0.264769  0.177447  0.005966 -0.184919  0.233605  0.287742  
119 -0.156343  0.258492  0.171485 -0.002076 -0.185321  0.

# 4. Model Creation

In [ ]:
y_train = train_df["Label"]
y_test = test_df["Label"]

## TFIDF Models

SVC adalah sebuah model Machine Learning yang digunakan untuk memisahkan data berdasarkan sebuah hyperplane. Hal ini bertujuan agar data-data dengan kategori atau kelas berbeda terpisahkan oleh sebuah celah yang sebesar mungkin.

Untuk parameter yang diubah, ada kernel, C, dan gamma. Kernel menjelaskan tipe pemisah hyperplane apa yang akan digunakan. C menjelaskan parameter penalti dari error. Bagian ini mengatur perbandingan antara pemisahan batas yang teratur dan pengelompokan data dengan benar. Terakhir, gamma menjelaskan parameter pengaturan untuk hyperplane selain linear, dimana nilai gamma yang makin tinggi, berarti fitting terhadap training data semakin dilakukan.

Sumber:

https://medium.com/all-things-ai/in-depth-parameter-tuning-for-svc-758215394769
https://medium.com/edureka/support-vector-machine-in-python-539dca55c26a

### SVC

In [ ]:
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.svm import SVC

svm = SVC()
params = {
    "C" : [0.01, 0.1, 1, 10, 100],
    "kernel" : ["linear", "poly", "rbf"],
    "gamma" : [0.001, 0.01, 0.1, 1]
}

grid_svm_tfidf = GridSearchCV(svm, param_grid = params)
grid_svm_tfidf.fit(TFIDF_train, y_train)


GridSearchCV(estimator=SVC(),
             param_grid={'C': [0.01, 0.1, 1, 10, 100],
                         'gamma': [0.001, 0.01, 0.1, 1],
                         'kernel': ['linear', 'poly', 'rbf']})

In [ ]:
## Model Prediction
pred = grid_svm_tfidf.predict(TFIDF_test)
eval_svm_tfidf = {
    'accuracy' : accuracy_score(y_test, pred),
    'precision' : precision_score(y_test, pred, average="weighted"),
    'recall' : recall_score(y_test, pred, average="weighted"),
    'F1' : f1_score(y_test, pred, average="weighted")
}

print(classification_report(y_test, pred))

               precision    recall  f1-score   support

     Business       0.80      1.00      0.89         4
Entertainment       0.62      1.00      0.77         5
     Politics       1.00      0.89      0.94         9
       Sports       1.00      0.50      0.67         6

     accuracy                           0.83        24
    macro avg       0.86      0.85      0.82        24
 weighted avg       0.89      0.83      0.83        24



In [ ]:
svm_tfidf_bestparams = grid_svm_tfidf.best_params_
svm_tfidf_bestparams

{'C': 10, 'gamma': 0.001, 'kernel': 'linear'}

In [ ]:
eval_svm_tfidf

{'accuracy': 0.8333333333333334,
 'precision': 0.8885416666666667,
 'recall': 0.8333333333333334,
 'F1': 0.8280124015418133}

### Random Forest

In [ ]:
from sklearn.ensemble import RandomForestClassifier

**Random Forest** adalah salah satu model klasifikasi machine learning dimana dasarnya adalah sebuah kumpulan decision tree. Kumpulan decision tree ini dibuat dengan metode 'Bagging' yang bertujuan untuk mengurangi kompleksitas model yang bisa overfit terhadap training data. Sama halnya dengan SVC, ada beberapa parameter yang diubah yaitu **n_estimators** yang menjelaskan jumlah tree yang digunakan untuk membuat forest. **max_depth** menjelaskan tentang seberapa dalam pohonnya, berarti jaalur terpanjang dari root sampai ke daunnya. **Criterion** ini adalah function yang mengukur kualitas dari split.

In [ ]:
rf = RandomForestClassifier()

params = {
    "n_estimators" : [100, 200],
    "criterion" : ["gini", "entropy", "log_loss"],
    "max_depth" : [None, 10, 20]
}

grid_rf_tfidf = GridSearchCV(rf, param_grid=params)
grid_rf_tfidf.fit(TFIDF_train, y_train)

GridSearchCV(estimator=RandomForestClassifier(),
             param_grid={'criterion': ['gini', 'entropy', 'log_loss'],
                         'max_depth': [None, 10, 20],
                         'n_estimators': [100, 200]})

In [ ]:
pred = grid_rf_tfidf.predict(TFIDF_test)
eval_rf_tfidf = {
    'accuracy' : accuracy_score(y_test, pred),
    'precision' : precision_score(y_test, pred, average="weighted"),
    'recall' : recall_score(y_test, pred, average="weighted"),
    'F1' : f1_score(y_test, pred, average="weighted")
}

print(classification_report(y_test, pred))

               precision    recall  f1-score   support

     Business       0.67      1.00      0.80         4
Entertainment       0.62      1.00      0.77         5
     Politics       0.86      0.67      0.75         9
       Sports       1.00      0.50      0.67         6

     accuracy                           0.75        24
    macro avg       0.79      0.79      0.75        24
 weighted avg       0.81      0.75      0.74        24



In [ ]:
rf_tfidf_bestparams = grid_rf_tfidf.best_params_
rf_tfidf_bestparams

{'criterion': 'gini', 'max_depth': 10, 'n_estimators': 200}

In [ ]:
eval_rf_tfidf

{'accuracy': 0.75,
 'precision': 0.8127480158730158,
 'recall': 0.75,
 'F1': 0.7415064102564103}

## Word2Vec - Skipgram

### SVC

In [ ]:
svm = SVC()
params = {
    "C" : [0.01, 0.1, 1, 10, 100],
    "kernel" : ["linear", "poly", "rbf"],
    "gamma" : [0.001, 0.01, 0.1, 1]
}

grid_svm_w2v = GridSearchCV(svm, param_grid = params)
grid_svm_w2v.fit(skipgram_train, y_train)

GridSearchCV(estimator=SVC(),
             param_grid={'C': [0.01, 0.1, 1, 10, 100],
                         'gamma': [0.001, 0.01, 0.1, 1],
                         'kernel': ['linear', 'poly', 'rbf']})

In [ ]:
pred = grid_svm_w2v.predict(skipgram_test)
eval_svm_w2v = {
    'accuracy' : accuracy_score(y_test, pred),
    'precision' : precision_score(y_test, pred, average="weighted"),
    'recall' : recall_score(y_test, pred, average="weighted"),
    'F1' : f1_score(y_test, pred, average="weighted")
}

print(classification_report(y_test, pred))

               precision    recall  f1-score   support

     Business       0.75      0.75      0.75         4
Entertainment       0.40      0.40      0.40         5
     Politics       0.82      1.00      0.90         9
       Sports       0.75      0.50      0.60         6

     accuracy                           0.71        24
    macro avg       0.68      0.66      0.66        24
 weighted avg       0.70      0.71      0.70        24



In [ ]:
svm_w2v_bestparams = grid_svm_w2v.best_params_
svm_w2v_bestparams

{'C': 100, 'gamma': 1, 'kernel': 'poly'}

In [ ]:
eval_svm_w2v

{'accuracy': 0.7083333333333334,
 'precision': 0.7026515151515151,
 'recall': 0.7083333333333334,
 'F1': 0.6958333333333333}

### Random Forest

In [ ]:
rf = RandomForestClassifier()

params = {
    "n_estimators" : [100, 200],
    "criterion" : ["gini", "entropy", "log_loss"],
    "max_depth" : [None, 10, 20]
}

grid_rf_w2v = GridSearchCV(rf, param_grid=params)
grid_rf_w2v.fit(skipgram_train, y_train)

GridSearchCV(estimator=RandomForestClassifier(),
             param_grid={'criterion': ['gini', 'entropy', 'log_loss'],
                         'max_depth': [None, 10, 20],
                         'n_estimators': [100, 200]})

In [ ]:
pred = grid_rf_w2v.predict(skipgram_test)
eval_rf_w2v = {
    'accuracy' : accuracy_score(y_test, pred),
    'precision' : precision_score(y_test, pred, average="weighted"),
    'recall' : recall_score(y_test, pred, average="weighted"),
    'F1' : f1_score(y_test, pred, average="weighted")
}

print(classification_report(y_test, pred))

               precision    recall  f1-score   support

     Business       0.75      0.75      0.75         4
Entertainment       0.40      0.40      0.40         5
     Politics       0.89      0.89      0.89         9
       Sports       0.50      0.50      0.50         6

     accuracy                           0.67        24
    macro avg       0.63      0.63      0.63        24
 weighted avg       0.67      0.67      0.67        24



In [ ]:
rf_w2v_bestparams = grid_rf_w2v.best_params_
rf_w2v_bestparams

{'criterion': 'gini', 'max_depth': 10, 'n_estimators': 100}

In [ ]:
eval_rf_w2v

{'accuracy': 0.6666666666666666,
 'precision': 0.6666666666666666,
 'recall': 0.6666666666666666,
 'F1': 0.6666666666666666}

# Model Comparison

 | Text Representation | ML Algorithm | ML Parameters | Accuracy | Precision | Recall | F1 Score |
| --- | --- | --- | --- | --- | --- | --- |
| TF-IDF | SVM | C = 10  <br /> gamma = 0.001 <br /> kernel = linear | 0.83 | 0.89 | 0.83 | 0.83 |
| TF-IDF | Random Forest | criterion = gini <br /> max_depth = 20 <br /> n_estimators = 200 | 0.75 | 0.81 | 0.75 | 0.74 |
| Word2Vec - Skipgram | SVM | C = 100  <br /> gamma = 1 <br /> kernel = poly | 0.71 | 0.70 | 0.71 | 0.70 |
| Word2Vec - Skipgram | Random Forest | criterion = entropy <br /> max_depth = 10 <br /> n_estimators = 100 | 0.67 | 0.67 | 0.67 | 0.67 |

Diatas adalah hasil dari performa setiap model yang dibuat. Terlihat bahwa model terbaik adalah model SVM yang menggunakan data dari representasi TF-IDF. Hal ini bisa terjadi karena jumlah data yang digunakan untuk kasus hanya sebatas 120, dimana jumlah ini masih terhitung sedikit untuk bisa membentuk sebuah korpus Word2Vec yang berisi data berguna.